### PR TITLE
Integrate WalletConnect to Dapp

### PR DESCRIPTION
**Pull request by nekhaly**
To support non Metmask users, WalletConnect support has been provided  
through this PR. This still lacks the proper UI and some minor issues  
are still around. The core idea here is two things, first is to  
implement `EthereumProvider` interface just like `MetamaskProvider` and  
then adapt the different functions with the slight differnces between  
Metamask and WalletConnect. The second thing, is proposing the store  
persistor as it was manadatory for keeping the state of the connected  
wallet (mainly for WalletConnect), so the user is kept connected even  
after reloading the screen. `pinia-plugin-persist` has been used for  
this goal, and in configuration, `localStorage` has been chosen as the  
storage technology (API) for browser. A new store `settings` has been  
added to keep the connection state and a unit test has been created  
for it.  
Testing should be added for the provider classes and the proper UI  
should be applied before merging this, including the disconnect from  
wallet feature. We have also and issue with showing the source chain  
value in select box after changing because window reloads and the value  
resets and is no longer available after that! A fix will be pushed  
later for this.

Resolves:  
#472

### CODE REVIEW

File: frontend/src/stores/settings.ts
Comment (weilbith) [inline frontend/src/stores/settings.ts:3]: I don't think need to define `Settings` externally here. There is only this store who cares about the type anyway. All sub-types that are used within the store might be interesting. But be aware that type inheritance also works out of the box.

File: frontend/src/components/RequestDialog.vue
Comment (manuelwedler) [inline frontend/src/components/RequestDialog.vue:150]: createWalletConnectProvider may return undefined. I think in this case we should not call this method.

File: frontend/src/stores/settings.ts
Comment (weilbith) [inline frontend/src/stores/settings.ts:11]: If I'm not mistaken this is not necessary. Pinia is reactive by default and it doesn't require to replace the whole property. In turn, I think you can get rid of the whole action and just set the property directly where needed. That is how Pinia is meant to work.

File: frontend/src/stores/settings.ts
Comment (weilbith) [inline frontend/src/stores/settings.ts:15]: Will become "useless" as soon as you flatten the state.

File: frontend/src/types/settings.ts
Comment (weilbith) [inline frontend/src/types/settings.ts:9]: Nitpick^10:

```suggestion
  MetaMask = 'metamask',
```

File: frontend/src/composables/useRpcUrls.ts
Comment (manuelwedler) [inline frontend/src/composables/useRpcUrls.ts:3]: This is not a composable. Composables encapsulate stateful logic and return some functions and/or references.
This looks like a helper function to me.

File: frontend/src/composables/useRpcUrls.ts
Comment (manuelwedler) [inline frontend/src/composables/useRpcUrls.ts:4]: Also the naming "list" seems not appropriate to me, because we are returning an object.

File: frontend/tests/unit/stores/settings.spec.ts
Comment (weilbith) [inline frontend/tests/unit/stores/settings.spec.ts:19]: I expect the action to become removed. So in this case this comment becomes useless. But anyway: you should check first if `connectedWallet` is undefined. Else the execution of `setConnectedWallet` is not correctly tested as this test does not implicitly set or verify that it was a different value before calling the function.

So either add a `expect(settings.settings?.connectedWallet).toBeUndefined()` or you do `settings.settings = { connectedWallet = undefined }`. The second has the advantage to make the test work even we should decide to (for what ever reason) have a different default state.

File: frontend/src/services/web3-provider/helper.ts
Comment (manuelwedler) [inline frontend/src/services/web3-provider/helper.ts:12]: same as the other comment: stores and composables should not be accessed in helper functions.

File: frontend/tests/unit/stores/settings.spec.ts
Comment (weilbith) [inline frontend/tests/unit/stores/settings.spec.ts:25]: Please try to avoid using other functions that are under test within the test of another one. This means if one is broken also the other test fails. Rather set the expected state manually via `settings.settings.connectedWallet = WalletType.MetaMask;`. Following this concept doesn't always work, but here it does and should be followed.

**Review by weilbith — CHANGES_REQUESTED** [review]
Thanks for all the time you put into this feature. It will be great to have WalletConnect in our application! 🙃 
There are a couple of comments, but many of them relate to each other. So it isn't as much as it might seam. After all I think we have a couple of options to improve our `EthereumProvider`. But that should not be done within this PR.

File: frontend/src/components/RequestDialog.vue
Comment (manuelwedler) [inline frontend/src/components/RequestDialog.vue:140]: I think we should show some feedback to the user if there is no MetaMask installed. But maybe that's already going to be addressed with the proper UI.

Also the button will not block / show the spinner until the requestSigner method is called. I think for async calls, it should block from the beginning.

**Review by manuelwedler — CHANGES_REQUESTED** [review]
Thank you for the work on this important feature! Quite nice to have this. There are couple of comments, but a lot of nitpicks. Many of them might already be raised by Thore.

File: frontend/src/components/RequestProcessing.vue
Comment (nekhaly) [reply frontend/src/components/RequestProcessing.vue:27]: I now see that I was in a rush enough to use the slang version of the language which ruins the grammar very well :D! Will go over this kind of stuff and fix it!

File: frontend/src/components/RequestDialog.vue
Comment (nekhaly) [reply frontend/src/components/RequestDialog.vue:6]: Yes, you are right... Actually, I was hoping for more than that so I didn't waste my time making it! I need a full wallet detailed box like that for Gnosis Safe app :)

File: frontend/src/components/RequestDialog.vue
Comment (nekhaly) [reply frontend/src/components/RequestDialog.vue:140]: I think in this certain case, yes... The assumption is the scenario of the user asks to connect through Metamask (user is not logged in), then when the box appears, the user doesn't login on Metamask, and then the user returns to the screen without assigning a provider, so he chooses again between WalletConnect and Metmask again!  
The problem I just realized is that the user has logged in but didn't connect to one of the accounts! Then your assumption is true about this case I think!

Comment (nekhaly) [reply frontend/src/components/RequestDialog.vue:140]: > I think we should show some feedback to the user if there is no MetaMask installed. But maybe that's already going to be addressed with the proper UI.
> 
I think this should be handled like Gnosis Safe app as well. We don't show the user anything if no Metamask on his/her browser, and if the user has decided to connect, we redirect him/her to the download page of Metamask!

> Also the button will not block / show the spinner until the requestSigner method is called. I think for async calls, it should block from the beginning.
Yup, nice caught.. I've seen it but totally forgot to fix it.. Thank you!

File: frontend/src/components/RequestDialog.vue
Comment (nekhaly) [reply frontend/src/components/RequestDialog.vue:150]: yes, right.. will do!

File: frontend/src/components/RequestDialog.vue
Comment (weilbith) [reply frontend/src/components/RequestDialog.vue:140]: Okay... So does that mean you will do it like I asked? 😬

Comment (nekhaly) [reply frontend/src/components/RequestDialog.vue:140]: The role applies to both approaches. I should think about a 3rd way :grimacing:!

File: frontend/tests/unit/stores/settings.spec.ts
Comment (nekhaly) [reply frontend/tests/unit/stores/settings.spec.ts:19]: It was already there and has been removed an idiot move! will add it again! Thank you!

File: frontend/tests/unit/stores/settings.spec.ts
Comment (nekhaly) [reply frontend/tests/unit/stores/settings.spec.ts:25]: That was a stupid assumption as well.. It seems I was asleep while implementing this because I assumed it will not reach this test case except passing the previous one!

File: frontend/src/types/settings.ts
Comment (nekhaly) [reply frontend/src/types/settings.ts:9]: Not a nitpick! I checked MetaMask lib and found that they use the camel-case version. I thought at the beginning that they deal with it as one word!

File: frontend/src/services/web3-provider/metamask-provider.ts
Comment (nekhaly) [reply frontend/src/services/web3-provider/metamask-provider.ts:105]: Would use the camel case in variable names but not for user interfacing text! but I agree with capitalizing the first letter!

Comment (nekhaly) [reply frontend/src/services/web3-provider/metamask-provider.ts:105]: OK.. Looks like I'm wrong again in my assumption.. even user interfacing text for MetaMask is camel-cased! will unify it over the app!

File: frontend/src/services/web3-provider/walletconnect-provider.ts
Comment (nekhaly) [reply frontend/src/services/web3-provider/walletconnect-provider.ts:94]: Yup! That's a good option! I was looking for a better design pattern to give more flexibility, but I think the one you suggest is enough so it's not complicated!

File: frontend/src/composables/useRequestFee.ts
Comment (nekhaly) [reply frontend/src/composables/useRequestFee.ts:18]: Yes, you are right! The issue was is that we were checking using the reference of requestManagerAddress which was always true.. We should have used requestManageAddress.value instead! Anyway, in my opinion, I'll remove requestManagerAddress check and use isSupportedChain because it gives a more meaning for the check is about IMO!

Comment (nekhaly) [reply frontend/src/composables/useRequestFee.ts:18]: And this would include for sure enhancing the helper functions themeselves!

Comment (nekhaly) [reply frontend/src/composables/useRequestFee.ts:18]: A generic fix for the whole composables issues is gonna be provided!

File: frontend/src/services/web3-provider/helper.ts
Comment (nekhaly) [reply frontend/src/services/web3-provider/helper.ts]: totally removed!

File: frontend/src/composables/useRequestFee.ts
Comment (nekhaly) [reply frontend/src/composables/useRequestFee.ts]: helper has been removed and code has been transferred into proper places (getters in stores or within `useWallet` composable)

File: frontend/src/services/web3-provider/walletconnect-provider.ts
Comment (nekhaly) [reply frontend/src/services/web3-provider/walletconnect-provider.ts:62]: I think is means if `walletConnectNetworkData === undefined || walletConnectNetworkData.name === 'unknown` then use the `name` from configuration!

**Comment by nekhaly** [comment]
Few comments that are unresolved, but it's ready to review right now IMO!

File: frontend/src/components/RequestDialog.vue
Comment (nekhaly) [reply frontend/src/components/RequestDialog.vue]: The spinner issue has been fixed!

**Comment by nekhaly** [comment]
Tested after rebase and works well! To the merge!

**Comment by weilbith** [comment]
The change requests by Manuel are still blocking the merge. As Manuel only works 1.5 days a week, we can't wait for that. So I'll use my permissions to force merge.

**Comment by istankovic** [comment]
:rocket: :fireworks: :wine_glass:

**Pull request merged**
